docs/zk_rollups_after_eip4844.md

```diff
@@ -0,0 +1,189 @@
+    Also verify that the provided commitment matches the provided versioned_hash.
+    """
+    # The data is encoded as follows: versioned_hash | z | y | commitment | proof | with z and y being padded 32 byte big endian values
+    assert len(input) == 192
+    versioned_hash = input[:32]
+    z = input[32:64]
+    y = input[64:96]
+    commitment = input[96:144]
+    proof = input[144:192]
+
+    # Verify commitment matches versioned_hash
+    assert kzg_to_versioned_hash(commitment) == versioned_hash
+
+    # Verify KZG proof with z and y in big endian format
+    assert verify_kzg_proof(commitment, z, y, proof)
+
+    # Return FIELD_ELEMENTS_PER_BLOB and BLS_MODULUS as padded 32 byte big endian values
+    return Bytes(U256(FIELD_ELEMENTS_PER_BLOB).to_be_bytes32() + U256(BLS_MODULUS).to_be_bytes32())
+
+
+def kzg_to_versioned_hash(commitment: KZGCommitment) -> VersionedHash:
+    return VERSIONED_HASH_VERSION_KZG + sha256(commitment)[1:]    
+```
+
+The `point_evaluation_precompile(versioned_hash, z, y, kzg_commitment, proof)` receives a versioned hash, the KZG commitment to the
+blob, and a KZG opening proof for point $z$ and value $y$ as input. It verifies that `kzg_commitment` corresponds to
+the `versioned_hash` provided and that the opening `proof` is valid (i.e. $f(z)
+=y$).
+
+## Blob Consistency Check
+
+Since rollup contract only can access to versioned hash instead of the raw transaction data (which was previously included
+as `calldata`), the main challenge we face is proving that $f$ indeed ==represents exactly the raw transaction data using circuits==.
+That is, we need to prove that our internal commitment scheme and KZG **commit to the same function** $f$ (can be called proof of
+equivalence).
+There is an easy approach in
+the case where the ZK rollup is BLS12-381 based, and a moderately harder approach for arbitrary ZK-SNARKs.
+
+### With BLS12-381 Modulus
+
+Given two polynomial commitments $C_1$ and $C_2$ over the same field (but not the same scheme, e.g. they could be KZG and FRI, or both
+KZG but with different trusted setup), here is the protocol:
+
+1. Let $x = hash(C_1, C_2)$ (based on [Fiat-Shamir heuristic](../terms/fiat_shamir.md))
+2. Compute $y = f(x)$
+3. Generate proof $\pi_1$, that proves $y = f(x)$ with respect to $C_1$
+4. Generate proof $\pi_2$, that proves $y = f(x)$ with respect to $C_2$
+
+The prover sends $C_1,C_2,y,\pi_1, \pi_2$. The verifier accepts if $y = f(hash(C_1,C_2))$ and the proofs $\pi_1$ and $\pi_2$ verify.
+This technique was summarized by
+Vitalik [here](https://ethresear.ch/t/easy-proof-of-equivalence-between-multiple-polynomial-commitment-schemes-to-the-same-data/8188).
+
+### With Any ZK-SNARKs
+
+We also choose $x$ as in previous section, and evaluate [barycentric equation](../terms/barycentric_equation.md) $f(x) = \dfrac{x^N}{N}
+\cdot \sum_{i}{\dfrac{d_i \cdot \omega^i}{x - \omega^i}}$ at our random $x$. This evaluation has to be done over BLS12-381 in our
```

Comment:
what is 'our circuit'?

Reply:
our own circuit

Reply:
where is this "our own circuit"? ;v

Reply:
the circuit that we define on our own that suits our needs (arbitrarily)